core/src/main/java/org/apache/iceberg/util/Tasks.java

```diff
@@ -212,7 +212,9 @@ public boolean run(Task<I, RuntimeException> task) {
           I item = iterator.next();
           try {
             runTaskWithRetry(task, item);
-            succeeded.add(item);
+            if (revertTask != null) {
+              succeeded.add(item);
+            }
```

Comment:
This is very suspicious to me. Can you explain why `Tasks` needs to change? It seems like this may be trying to handle the case where a revert happens later and so the success didn't necessarily go through. But this check doesn't look like the right way to accomplish that.

Reply:
Let me explain why the change is needed here:  
At present, when rewriting, it will be rewritten with the `RewriteFileGroup` as the rewrite unit through the `Tasks` tool class without `revertWith` method:  
    // Start rewrite tasks
    Tasks.foreach(groupStream)
        .suppressFailureWhenFinished()
        .executeWith(rewriteService)
        .noRetry()
        .onFailure((fileGroup, exception) -> LOG.error("Failure during rewrite group {}", fileGroup.info(), exception))
        .run(fileGroup -> commitService.offer(rewriteFiles(ctx, fileGroup)));  
But at this time, after the rewrite is finished, there is no need to revert the successful task. As you can see from the following code, unless we call and set `revertTask`, the logic here will not be executed anyway:  
    public Builder<I> revertWith(Task<I, ?> task) {
      this.revertTask = task;
      return this;
    }  
    ...  
    if (taskFailed.get() && revertTask != null) {
      // at least one task failed, revert any that succeeded
      for (final I item : succeeded) {
        futures.add(service.submit(new Runnable() {
          @Override
          public void run() {
            if (stopRevertsOnFailure && revertFailed.get()) {
              return;
            }


            boolean failed = true;
            try {
              revertTask.run(item);
              failed = false;
            } catch (Exception e) {
              LOG.error("Failed to revert task", e);
              // swallow the exception
            } finally {
              if (failed) {
                revertFailed.set(true);
              }
            }
          }
        }));
      }  
But the current logic is that no matter what, the reference of the successful task will be saved to the concurrent queue of succeeded, and the `RewriteFileGroup` that has been rewritten will also be saved. If we want to release the memory of `RewriteFileGroup`, then we should also remove unnecessary references to it here